### DISCUSSION TITLE
Understanding Host/Docker Networking

### DISCUSSION THREAD

**Discussion by Nostang3** [Jip-Hop/jailmaker#137] [General]
The installation guide suggests not to use the host network for docker. And by default it has access to the host network. I would like a little more clarification on things. 

* What happens when I remove `docker_compatible=1` from the config? I had the jail setup using that option. Will it continue to work with no side effects? Do I need to create a new jail without it using the docker config then move everything over?
* If, inside the jail, I create a bridge with docker will this isolate it from the host network? If not then I have to create a bridge on Truenas? Is the host bridge option still acceptable if you have a static ip for the host and only want to connect to the docker containers via http://host_ip:port? If not what is the best way for that?

**Comment by Lockszmith-GH** [comment]
Using the latest version you should not use `docker_compatible=1` as it is being deprecated.

The updated [README.md Docker section](https://github.com/Jip-Hop/jailmaker?tab=readme-ov-file#docker) now states the reasons.

Keep in mind this all has to do with linux kernel architecture and the implementation of [systemd-nspawn networking](https://manpages.debian.org/bookworm/systemd-container/systemd-nspawn.1.en.html#Networking_Options).

**Comment by Nostang3** [comment]
I've read that but I still made this discussion asking for more verbose answers because the documents assume you have a lot of knowledge about things at least to me. You're told to do things but not why. Like `But keep in mind that you need to add --system-call-filter='add_key keyctl bpf'` but doesn't explain what that line does. I want to make sure that I'm separating my docker network from my host network in regards to the ip talbes but not severe the connection completely while also maintaining the http://host-static-ip:port external access.

**Reply by Jip-Hop** [reply]
You may read what the `--system-call-filter=` options does in the nspawn docs and arch linux wiki, both [linked from References the main readme](https://github.com/Jip-Hop/jailmaker?tab=readme-ov-file#references). It's not an option from jailmaker itself, just passed through to systemd-nspawn so you need to refer to their manual.

Regarding migrating away from the `docker_compatible` I recommend you create a new jail for docker by using the [docker config template](https://github.com/Jip-Hop/jailmaker/blob/main/templates/docker/README.md). Use macvlan networking and for the time being don't add any bind mounts. You should get a dynamic IP address for this jail via DHCP. You can make this a fixed IP address in your router (easy) or refer to [the Advanced Networking docs for inspiration on how to setup a static IP address using config files](https://github.com/Jip-Hop/jailmaker/blob/main/NETWORKING.md).

Then you may spin up a little test container, e.g. a webserver, and access it from http://jail-static-ip:port. Once you're comfortable with this setup I recommend you stop your current docker jail, add the bind mounts you used to have to your new docker jail and restart your new docker jail. Then run your containers from that jail (and never start your old jail again, you don't want to accidentally have the same containers, with the same bind mounts, running simultaneously from 2 jails). If you use hostnames for the services you host with docker you'll have to update them (e.g. in your router or DNS server), since your services are now hosted on jail-static-ip instead of host_ip.

**Reply by Lockszmith-GH** [reply]
@Nostang3, you are right in wanting to understand more, it's just that that 'more' is about the Linux kernel architecture and it's features of virtualization/containarization are utilized by `systemd-nspawn`, and those details are documented way better elseware.

[add_key](https://linux.die.net/man/2/add_key), [keyctl](https://linux.die.net/man/2/keyctl) and [bpf](https://www.kernel.org/doc/html/latest/networking/filter.html) are all Linux kernel mechanism which docker utilizes inside the jail, and need the ability to control.

**Reply by Nostang3** [reply]
> Regarding migrating away from the `docker_compatible` I recommend you create a new jail for docker by using the [docker config template](https://github.com/Jip-Hop/jailmaker/blob/main/templates/docker/README.md). Use macvlan networking and for the time being don't add any bind mounts. You should get a dynamic IP address for this jail via DHCP. You can make this a fixed IP address in your router (easy) or refer to [the Advanced Networking docs for inspiration on how to setup a static IP address using config files](https://github.com/Jip-Hop/jailmaker/blob/main/NETWORKING.md).
> 
> Then you may spin up a little test container, e.g. a webserver, and access it from http://jail-static-ip:port. Once you're comfortable with this setup I recommend you stop your current docker jail, add the bind mounts you used to have to your new docker jail and restart your new docker jail. Then run your containers from that jail (and never start your old jail again, you don't want to accidentally have the same containers, with the same bind mounts, running simultaneously from 2 jails). If you use hostnames for the services you host with docker you'll have to update them (e.g. in your router or DNS server), since your services are now hosted on jail-static-ip instead of host_ip.

Thanks, I'll give that at try.
